### PR TITLE
Maxime/elevator simple collision check

### PR DESCRIPTION
**Pull request by Franchovy**
Modified elevator code so that:

- Activate sets a child sprite and sets movement to make
- If no activation, elevator snaps to closest tile if within `tileAdjustmentPx`
- Movement code automatically handles own and child collisions. Prevents movement if elevator or child collides (with something else but themselves).

We'll need to make some adjustments if we want to handle multiple sprites in the collision. But for the playtester version (only player) this seems to work for me.

@CMorooney go hard in your review! Point out anything that doesn't make sense or that could be improved.

### CODE REVIEW

File: src/sprites/Elevator.lua [hunk 1]
```diff
@@ -72,118 +66,225 @@ function Elevator:collisionResponse(_)
...
+    if key == KEYNAMES.Down then
+      return self.speed
+    elseif key == KEYNAMES.Up then
+      return -self.speed
+    end
+  end
 
-  return math.min(displacementRemaining, self.speed) * (isInverseDirection and -1 or 1)
-end
+  -- If key is not a correct activation key, return 0
```
Comment (CMorooney) [inline src/sprites/Elevator.lua]: should we consider asserting more strongly? (there are a few fallbacks like this, same question for all of them)

File: src/sprites/Elevator.lua [hunk 2]
```diff
@@ -72,118 +66,225 @@ function Elevator:collisionResponse(_)
...
+    -- If collision happens to elevator above, skip.
+
+    local shouldSkipCollision = 
+      (self == spriteToCheck and collision.normal.y == -1) or
+      (self ~= spriteToCheck and collision.normal.y == 1)
+    
+    if shouldSkipCollision then
+      -- Block collision
+      isCollisionCheckPassed = false
+
```
Comment (CMorooney) [inline src/sprites/Elevator.lua]: spaces/empty line

File: src/sprites/Elevator.lua [hunk 3]
```diff
@@ -72,118 +66,225 @@ function Elevator:collisionResponse(_)
...
+  local actualSpriteToCheckX, actualSpriteToCheckY, collisions = spriteToCheck:checkCollisions(idealSpriteToCheckX, idealSpriteToCheckY)
+  local isCollisionCheckPassed = true
+
+  for _, collision in pairs(collisions) do
+    -- If collision happens to elevator above, skip.
+
+    local shouldSkipCollision = 
+      (self == spriteToCheck and collision.normal.y == -1) or
+      (self ~= spriteToCheck and collision.normal.y == 1)
+
```
Comment (CMorooney) [inline src/sprites/Elevator.lua]: spaces in empty line

File: src/sprites/Elevator.lua [hunk 4]
```diff
@@ -72,118 +66,225 @@ function Elevator:collisionResponse(_)
...
-  
-  -- Set movement update scalar
-  self.movement = movement
+end
 
-  self.displacement += self.movement * _G.delta_time
+--- Checks collision for frame, also checking if child collides. Returns a partial movement for itself
+--- if elevator or child collides with another object.
+local function checkIfCollides(self, idealX, idealY, spriteToCheck)
+  local spriteToCheck = spriteToCheck or self
```
Comment (CMorooney) [inline src/sprites/Elevator.lua:110]: could we use a different name for the arg and the local variable

File: src/sprites/Elevator.lua [hunk 5]
```diff
@@ -72,118 +66,225 @@ function Elevator:collisionResponse(_)
...
-      self.movement = 0
+  -- Check if elevator collides
+
+  local actualSpriteToCheckX, actualSpriteToCheckY, collisions = spriteToCheck:checkCollisions(idealSpriteToCheckX, idealSpriteToCheckY)
+  local isCollisionCheckPassed = true
+
+  for _, collision in pairs(collisions) do
+    -- If collision happens to elevator above, skip.
+
+    local shouldSkipCollision =
```
Comment (CMorooney) [inline src/sprites/Elevator.lua]: trailing space

File: src/sprites/Elevator.lua [hunk 6]
```diff
@@ -72,118 +66,225 @@ function Elevator:collisionResponse(_)
...
+  local idealSpriteToCheckY = idealY - self.y + spriteToCheck.y
 
-  -- If close to start or end, adjust displacement & cancel movement
-  if self.displacement - self.displacementStart < 1 or
-    self.displacementEnd - self.displacement < 1 then
-      self:displacementAdjustToTile()
-      self.movement = 0
+  -- Check if elevator collides
+
+  local actualSpriteToCheckX, actualSpriteToCheckY, collisions = spriteToCheck:checkCollisions(idealSpriteToCheckX, idealSpriteToCheckY)
```
Comment (CMorooney) [inline src/sprites/Elevator.lua:116]: it's too bad we have to run collision checks again, ostensibly the sprite is already making those during any RB updates. Probably isn't going to slow us down in practice too much knowing the game, just a thought

File: src/sprites/Elevator.lua [hunk 7]
```diff
@@ -72,118 +66,225 @@ function Elevator:collisionResponse(_)
...
+  -- Get new position using displacement
+
+  local x, y = getPositionFromDisplacement(self, self.displacement + movement)
+
+  -- Check collisions for self
+  
+  local isCollisionCheckPassed, x, y = checkIfCollides(self, x, y)
+
+  -- Check collisions for child
+
```
Comment (CMorooney) [inline src/sprites/Elevator.lua:194]: empty line with spaces

File: src/sprites/Elevator.lua [hunk 8]
```diff
@@ -72,118 +66,225 @@ function Elevator:collisionResponse(_)
...
+    local centerX = x + self.width / 2
+
+    local offsetY = 0
+    if movement > 0 and self.fields.orientation == ORIENTATION.Vertical then
+      -- For moving down, move player slightly into elevator for better collision activation
+      offsetY = 2
+    end
+
+    self.spriteChild:moveTo(
+      centerX - self.spriteChild.width / 2,
```
Comment (CMorooney) [inline src/sprites/Elevator.lua]: trailing space

File: src/sprites/Elevator.lua [hunk 9]
```diff
@@ -72,118 +66,225 @@ function Elevator:collisionResponse(_)
...
+    if movement > 0 and self.fields.orientation == ORIENTATION.Vertical then
+      -- For moving down, move player slightly into elevator for better collision activation
+      offsetY = 2
+    end
+
+    self.spriteChild:moveTo(
+      centerX - self.spriteChild.width / 2, 
+      y - self.spriteChild.height + offsetY
+    )
+  end
```
Comment (CMorooney) [inline src/sprites/Elevator.lua]: trailing space

File: src/sprites/Elevator.lua [hunk 10]
```diff
@@ -72,118 +66,225 @@ function Elevator:collisionResponse(_)
...
+--
 
-  -- Skip displacement to the nearest tile 
+--- Sets movement to be executed in the next update() call using vector.
+--- *param* key - the player input key direction (KEYNAMES)
+--- *returns* the distance covered in the activation.
+function Elevator:activate(sprite, key)
+  -- Gets applied movement using key, self.speed and self.orientation
+  local activationMovement = getActivationMovement(self, key)
+
```
Comment (CMorooney) [inline src/sprites/Elevator.lua]: empty line with spaces

File: src/sprites/Elevator.lua [hunk 11]
```diff
@@ -72,118 +66,225 @@ function Elevator:collisionResponse(_)
...
+  -- If activated, set update variables for movement
+  if activationMovement ~= 0 then
+    -- Set child sprite for collision check
+    self.spriteChild = sprite
 
-    self:updatePosition()
+    -- Set movement update scalar
+    self.movement = activationMovement
+  end
+
```
Comment (CMorooney) [inline src/sprites/Elevator.lua]: empty line with spaces

File: src/sprites/Elevator.lua [hunk 12]
```diff
@@ -72,118 +66,225 @@ function Elevator:collisionResponse(_)
...
-  elseif self.fields.orientation == ORIENTATION.Vertical then
+  self.isCollisionsDisabledForFrame = true
+end
+
+function Elevator:update()
+  Elevator.super.update(self)
+
+  -- Get if elevator has been activated
+  local movement = self.movement
+
```
Comment (CMorooney) [inline src/sprites/Elevator.lua]: empty line with spaces

File: src/sprites/Elevator.lua [hunk 13]
```diff
@@ -193,6 +294,11 @@ function Elevator:update()
 
     self.isCollisionsDisabledForFrame = false
   end
+
+  -- Reset update variables
+
```
Comment (CMorooney) [inline src/sprites/Elevator.lua]: empty line with spaces

**Review by CMorooney — APPROVED** [review]
looking good! played around in the test level a bit wasn't able to get stuck or anything

File: src/sprites/Elevator.lua (hunk 1 above)
Comment (Franchovy) [reply src/sprites/Elevator.lua]: Need to double-check that Player isn't pushing random other keys to the elevator. I wouldn't have considered it an error necessarily that the elevator receives other keys but for this version you have a good point.

File: src/sprites/Elevator.lua (hunk 2 above)
Comment (Franchovy) [reply src/sprites/Elevator.lua]: Need to update my linter :D

File: src/sprites/Elevator.lua (hunk 4 above)
Comment (Franchovy) [reply src/sprites/Elevator.lua:110]: This was done to allow for recursion. Also it's common in lua to "or" an argument i.e. provide a default value.

File: src/sprites/Elevator.lua (hunk 6 above)
Comment (Franchovy) [reply src/sprites/Elevator.lua:116]: This is a very valid point. For now I don't have a better idea... the PD collision system can't handle two moving sprites moving into each other, so we'd have to implement our own collision resolving potentially. Agree that RB can be the source of this improvement.

File: src/sprites/Elevator.lua (hunk 4 above)
Comment (Franchovy) [reply src/sprites/Elevator.lua:110]: Added in the recursion :D

File: src/sprites/Elevator.lua (hunk 2 above)
Comment (Franchovy) [reply src/sprites/Elevator.lua]: Updated my linter ;)

File: src/sprites/Elevator.lua (hunk 1 above)
Comment (Franchovy) [reply src/sprites/Elevator.lua]: Player is checking random keys, so added the assert :)

**Pull request merged**